src/compiler/crystal/syntax/parser.cr

```diff
@@ -4033,6 +4036,12 @@ module Crystal
       end
     end
 
+    def check_valid_param_name(param_name)
+      if param_name.ends_with?(/\?|!/)
+        raise "invalid param name"
+      end
```

Comment:
Small optimization (no regex) + rephrased error message to make it more clear about the name in question:

```suggestion
      if param_name[-1].in?('?', '!')
        raise "invalid parameter name: #{param_name}"
      end
```

Reply:
When I use this version I get a bunch of `Index out of bounds (IndexError)` errors.

I guess I should first check if the `param_name` string is not empty, what do you think?

Reply:
You can go with this:

```suggestion
      if param_name.ends_with?("?") || param_name.ends_with?("!")
        raise "invalid param name"
      end
```

Reply:
I've edited the suggestion

Reply:
Ok, so the choice is between:

```
param_name[-1]?.in?('?', '!')
```

```
!param_name.empty? && param_name[-1].in?('?', '!')
```

```
param_name.ends_with?("?") || param_name.ends_with?("!")
```

I think the last one is the most clear.

Is there an advantage in using version 1 or 2?
Which one is the most idiomatic?